### DISCUSSION TITLE
COPY FROM a Pandas DataFrame Example Issue

### DISCUSSION THREAD

**Discussion by cmcclung-ANGENG** [kuzudb/kuzu#3499] [Q&A] [answered]
Hi,

I was trying to  run the example copy from Pandas.  But the example is missing an ending bracket.

So it's:
```py
conn.execute("COPY Person FROM (LOAD FROM df WHERE age < 30 RETURN *)")
```

But why can't I pass in the Pandas Dataframe as a parameter? I get different errors like `Unknown parameter type <class 'pandas.core.frame.DataFrame'>`.

On other platforms like Java what would the equivalent Subquery be?

Curious what I can pass in as I'd be importing from different sources into a single db schema.

**Comment by prrao87** [comment, marked as answer]
Thanks for pointing it out, we'll get that example fixed right away. To your second point, the scanning from Pandas is designed to be for Python users. Kùzu is able to scan Pandas DataFrames both the numpy and Arrow backends by using a [memcpy](https://cplusplus.com/reference/cstring/memcpy/) operation, allowing it access to the raw pointers that hold the data. So there's no need to pass it in as a parameter - you'd just access the underlying object via the Python API.

Are you trying to access a specific format from your Java application code? Not sure how you'd intend to use Pandas alongside Java 😅.

**Reply by cmcclung-ANGENG** [reply]
Ah, that makes sense.

No, not using Pandas alongside Java ;).  I got confused with the way the parameters were handled.

In the other languages I'd be building up raw types from the likes of yaml, db etc.  Then insert the normalised version.

**Reply by prrao87** [reply]
Makes sense :). You can think of using the `LOAD FROM` subquery after you convert the raw data into a format that's supported (CSV, Parquet), or, you could alternatively ingest the data into node/relationship tables one record-at-a-time by using the `CREATE` or `MERGE` statements. Hope that helps!
